packages/vite/src/node/plugins/resolve.ts

```diff
@@ -770,12 +773,32 @@ function packageEntryFailure(id: string, details?: string) {
   )
 }
 
+const ssrConditions: string[] = []
+process.execArgv.forEach((arg, i, argv) => {
+  const [key, value] = arg.includes('=')
+    ? arg.split('=')
+    : argv.slice(i - 1, i + 1)
+
+  if (key === '-C' || key === '--conditions') {
+    ssrConditions.push(value)
+  }
+})
+
 function resolveExports(
   pkg: PackageData['data'],
   key: string,
   options: InternalResolveOptions,
-  targetWeb: boolean
+  targetWeb: boolean,
+  ssr?: boolean
 ) {
+  if (ssr && !options.isBuild) {
+    // To ensure SSR imports are resolved identically regardless of whether
+    // the importer is externalized or not, the conditions need to match what
+    // the Node.js runtime is using.
+    return _resolveExports(pkg, key, {
+      conditions: ssrConditions
+    })
+  }
```

Comment:
Is it fine that we're running this for SSR dev only? Otherwise when we run an SSR build and bundle a node package, it would skip this part, which I think causes inconsistencies in dev mode.

Reply:
We can't know at build time which `--conditions` flags will be used when running the SSR build. :\

Reply:
Oh true that's tricky 🤔 Perhaps we should only apply this code for externalized packages in dev mode only? Since in SSR build, non-externalized packages will use Vite's resolve options. That way we can reduce the inconsistencies. Seems like you've commented against this behaviour in-code though, what do you think?

Reply:
To ensure resolution is consistent (externalized or not) in SSR builds, we would need to add a `ssr.conditions` option and add a runtime warning when the `--conditions` flags don't match it. Or there might be a way to inject `--conditions` flags at runtime automatically? Idk.

Reply:
Maybe the best option is to (as @brillout as suggested before) avoid Vite's resolution algo for imports of externalized packages **from non-externalized modules**. If we do that, then it won't matter if there's inconsistencies between Node's externalized resolution and Vite's bundled resolution, since they (practically speaking) never resolve to the same package (and so mismatched `conditions` have no confusing effects).

Reply:
> Maybe the best option is (as @brillout has suggested before) to avoid Vite's resolution algo for imports of externalized packages **from non-externalized modules as well**

Yeah this sounds like a fine plan too. I don't think non-externalized modules should rely on `conditions` though since that usually means the user explicitly wants to bundle them using Vite's resolve terms. For externalized modules, we can leave it as is so the environment platform resolves it themselves.